File: self pased DSA/Tree/11-Continuous-Subarray-Sum.cpp
/**
 * Given an integer array nums and an integer k, return true if nums has a good subarray or false otherwise.

A good subarray is a subarray where:

its length is at least two, and
the sum of the elements of the subarray is a multiple of k.
Note that:

A subarray is a contiguous part of the array.
An integer x is a multiple of k if there exists an integer n such that x = n * k. 0 is always a multiple of k.
 */


 class Solution {
public:
    /* 
    approach: the approach is simple we calculate prefix Sum and check if reminder of prefixSum % k is already present 
    and its index is not same than we have a subarray sum which is divisible by k
     Example: 
     [a, b, c, d, e, f, ... ]
     lets say (c+d+e) is divisible by k
     sum1 = a+b, sum2 = a+b+c+d+e
     then (sum2-sum1) = c+d+e, which is divisible by k
     => so we can write (sum2-sum2)%k==0
     => sum2%k == sum1%k
     => we can say if remider divided by k for sum1 and sum2 is equal than we have a contiguous subarray sum which could be divisible by k
     */
    
    bool checkSubarraySum(vector<int>& nums, int k) {
        unordered_map<int,int> mp;
        // coorner case: if divider is zero i.e. k=0 then check 2 consecutive zeros 
        if(k==0){
            for(int i=1; i<nums.size(); i++){
                if(nums[i]==0 && nums[i-1]==0)  
                    return true;
            }
            return false;
        }


        long long sum = 0;
        mp.insert({0, -1}); // insert reminder 0 have -1 index this will work for prefixhum zero reminder
        //  in case if you get prefixSum % k==0 then check if its length is greater than 1
        // then 
        for(int i=0; i<nums.size(); i++){
            sum += nums[i]; // prefix sum
            int rem = sum%k; // reminder of prefix sum

            if(mp.find(rem)!=mp.end() && i-mp[rem]>1){
                return true;
            }
            mp.insert({rem, i});
        }
        return false;
    }
};

File: code_java/self pased DSA/Tree/11-Continuous-Subarray-Sum.java
/**
 * Problem Summary:
 * Given an integer array nums and an integer k, determine if there exists a "good" subarray.
 * A "good" subarray has length at least 2 and its sum is a multiple of k (including zero).
 *
 * Approach:
 * - Use prefix sums and modulo k to identify subarrays with sums divisible by k.
 * - Store the earliest index of each prefix sum remainder in a HashMap.
 * - If the same remainder appears again and the distance between indices is > 1, return true.
 * - Handle special case when k == 0 by checking for consecutive zeros.
 *
 * Why it works:
 * - If prefixSum[j] % k == prefixSum[i] % k, then sum of subarray (i+1 to j) is divisible by k.
 * - HashMap allows O(1) checks for repeated remainders.
 * - Checking distance > 1 ensures subarray length >= 2.
 */
import java.util.HashMap;

public class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        // Special case: k == 0, check for at least two consecutive zeros
        if (k == 0) {
            for (int i = 1; i < nums.length; i++) {
                if (nums[i] == 0 && nums[i - 1] == 0) {
                    return true;
                }
            }
            return false;
        }

        HashMap<Integer, Integer> remainderIndexMap = new HashMap<>();
        remainderIndexMap.put(0, -1); // remainder 0 at index -1 to handle subarrays starting at index 0

        long sum = 0;
        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];
            int remainder = (int)(sum % k);

            if (remainderIndexMap.containsKey(remainder)) {
                int prevIndex = remainderIndexMap.get(remainder);
                // Check subarray length is at least 2
                if (i - prevIndex > 1) {
                    return true;
                }
            } else {
                remainderIndexMap.put(remainder, i);
            }
        }

        return false;
    }
}